src/validation.cpp

```diff
@@ -2633,6 +2766,322 @@ static unsigned int GetBlockScriptFlags(const CBlockIndex* pindex, const Consens
     return flags;
 }
 
+bool ConfirmAllPreDaedalusAddresses(
+        CValidationState& state,
+        const Consensus::Params& consensus,
+        const CBlockIndex* pindexPrev)
+{
+    const auto height = pindexPrev->nHeight + 1;
+
+    //Don't do the indexing if we are not on the first block during the daedalus deployment.
+    if(height < consensus.vDeployments[Consensus::DEPLOYMENT_DAEDALUS].start_block) { 
+        return true;
+    }
+
+    if(height > consensus.vDeployments[Consensus::DEPLOYMENT_DAEDALUS].start_block) { 
+        return prefviewdb->AreAllPreDaedalusAddressesConfirmed();
+    }
+
+    //One time confirmation of all addresses before the daedalus block
+    prefviewdb->ConfirmAllPreDaedalusAddresses();
+}
+
+bool ConfirmAddresses(const CBlock& block)
+{
+    ReferralSet referrals_in_block;
+    BuildReferralSet(block, referrals_in_block);
+
+    for(const auto& tx : block.invites) {
+        assert(tx);
+
+        for (int i = 0; i < tx->vout.size(); i++) {
+            const auto& out = tx->vout[i];
+            const auto address = ExtractAddress(out);
+            if(address.second == 0) {
+                if(out.nValue == 0) continue;
+                else return false;
+            }
+
+            const referral::Referral* ref_to_confirm = nullptr;
+            auto maybe_referral = prefviewdb->GetReferral(address.first);
+
+            if(!maybe_referral) {
+                if(referrals_in_block.count(address.first)) {
+                    auto ref = std::find_if(
+                            block.m_vRef.begin(), block.m_vRef.end(), 
+                            [&address](const referral::ReferralRef ref) {
+                                return ref->GetAddress() == address.first;
+                            });
+                    if(ref != block.m_vRef.end()) {
+                        ref_to_confirm = ref->get();
+                    }
+                }
+            } else {
+                ref_to_confirm = maybe_referral.get_ptr();
+            }
+
+            if(ref_to_confirm == nullptr) {
+                return false;
+            }
+
+            if(!prefviewdb->ConfirmReferral(*ref_to_confirm, *tx, i)) {
+                return false;
+            }
+        }
+    }
+
+    return true;
+}
+
+bool ValidateAddressesAreConfirmed(const CBlock& block)
+{
+    for(const auto& tx : block.vtx) {
+        assert(tx);
+
+        for (const auto& out : tx->vout) {
+            const auto address = ExtractAddress(out);
+            if(address.second == 0) {
+                if(out.nValue == 0) continue;
+                else return false;
+            }
+
+            if(!prefviewdb->IsConfirmed(address.first)) {
```

Comment:
so, we do not allow to use addresses for transactions that were confirmed in the same block?